src/porepy/numerics/time_step_control.py

```diff
@@ -0,0 +1,273 @@
+from typing import Tuple, Union
+
+__all__ = ["TimeSteppingControl"]
+
+
+class TimeSteppingControl:
+    """Parent class for iteration-based time stepping control routine."""
+
+    def __init__(
+        self,
+        schedule: list,
+        dt_init: float,
+        dt_min_max: Tuple[float, float],
+        iter_max: int = 15,
+        iter_optimal_range: Tuple[int, int] = (4, 7),
+        iter_lowupp_factor: Tuple[float, float] = (1.3, 0.7),
+        recomp_factor: float = 0.5,
+        recomp_max: int = 10,
+        print_info: bool = False,
+    ):
+        """Computes the next time step based on the number of non-linear iterations.
+
+        Parameters:
+            schedule (List): List containing the target times for the simulation.
+                The time-stepping algorithm will adapt the time step so that the target
+                times are guaranteed to be hit/reached. The list must contain minimally
+                two elements, corresponding to the the initial and final simulation times.
+                Lists of length > 2 must contain strictly increasing times.
+                Examples of valid inputs are:
+                  [0, 1], [0, 10, 30, 50], [0, 1*pp.HOUR, 3*pp.HOUR].
+                Examples of invalid inputs are:
+                  [1], [1, 0], [0, 1, 1, 2].
+            dt_init (float): Initial time step.
+            dt_min_max (Tuple of float): Minimum and maximum permissible time steps.
+            iter_max (int): Maximum number of iterations. Default is 15.
+            iter_optimal_range (Tuple of int): Lower and upper optimal iteration range.
+                Default is (4, 7).
+            iter_lowupp_factor (Tuple of float, optional): Lower and upper multiplication
+                factors. Default is (1.3, 0.7).
+            recomp_factor (float). Failed-to-converge recomputation factor.
+                Default is 0.5.
+            recomp_max (int). Failed-to-converge maximum recomputation attempts.
+                Default is 10.
+            print_info (bool). Print time-stepping information. Default is False.
+
+        """
+
+        # Sanity checks on input paramters
+        if len(schedule) < 2:
+            s = "Schedule list must have at least two items, representing the initial and"
+            s += " final simualtion time."
+            raise ValueError(s)
+        elif schedule[0] < 0:
```

Comment:
Is this needed?

Reply:
The idea here is to catch negative times. But yes, I've modified this check to raise an error if `any` item in schedule is negative, not only the first one.